src/plots/cartesian/align_period.js

```diff
@@ -0,0 +1,113 @@
+/**
+* Copyright 2012-2020, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var isNumeric = require('fast-isnumeric');
+var Lib = require('../../lib');
+var ms2DateTime = Lib.ms2DateTime;
+var dateTime2ms = Lib.dateTime2ms;
+var constants = require('../../constants/numerical');
+var ONEDAY = constants.ONEDAY;
+var ONEAVGMONTH = constants.ONEAVGMONTH;
+var ONEAVGYEAR = constants.ONEAVGYEAR;
+
+module.exports = function alignPeriod(trace, ax, axLetter, vals) {
+    if(ax.type !== 'date') return vals;
+
+    var alignment = trace[axLetter + 'periodalignment'];
+    if(!alignment) return vals;
+
+    var period = trace[axLetter + 'period'];
+    if(isNumeric(period)) {
+        period = +period;
+        if(period <= 0) return vals;
+    } else if(typeof period === 'string' && period.charAt(0) === 'M') {
+        var n = +(period.substring(1));
+        if(n > 0 && Math.round(n) === n) period = n * ONEAVGMONTH;
+        else return vals;
+    }
+
+    if(period > 0) {
+        var calendar = ax.calendar;
+
+        var isStart = 'start' === alignment;
+        // var isMiddle = 'middle' === alignment;
+        var isEnd = 'end' === alignment;
+
+        var offset = (new Date()).getTimezoneOffset() * 60000;
```

Comment:
Can we avoid `offset` if we use all UTC variants of the date accessors? ie `getUTCFullYear` instead of `getFullYear` etc? I expect as is this will have DST problems.

Reply:
Good call. Done in https://github.com/plotly/plotly.js/pull/5074/commits/be5228176e54f0e068dd155ee6c052351b0c02d5.